File: Learn_Android/XGNotification/app/src/main/java/com/example/user/xgnotification/MainActivity.java
package com.example.user.xgnotification;

import android.support.v7.app.AppCompatActivity;
import android.os.Bundle;
import android.util.Log;
import android.widget.Button;

import com.tencent.android.tpush.XGIOperateCallback;
import com.tencent.android.tpush.XGPushConfig;
import com.tencent.android.tpush.XGPushManager;

public class MainActivity extends AppCompatActivity {

    Button button;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);
//        button=(Button)findViewById(R.id.sned_btn);
        XGPushConfig.enableDebug(this,true);
        XGPushManager.registerPush(this, new XGIOperateCallback() {
            @Override
            public void onSuccess(Object o, int i) {
                    Log.d("TPush","注册成功，设备token为"+o);
                }

            @Override
            public void onFail(Object o, int i, String s) {
                    Log.d("Tpush","注册失败，错误码"+i+"，错误信息"+s);
            }
        });



    }
}

File: Learn_Android/Di_Yi_Hang_Dai_Ma(3rd edition)/JetpackTest/app/src/main/java/com/example/jetpacktest/MainActivity.kt
package com.example.jetpacktest

import android.content.SharedPreferences
import androidx.appcompat.app.AppCompatActivity
import android.os.Bundle
import android.widget.Button
import android.widget.TextView
import androidx.core.content.edit
import androidx.lifecycle.Observer
import androidx.lifecycle.ViewModelProvider
import androidx.work.BackoffPolicy
import androidx.work.OneTimeWorkRequest
import androidx.work.WorkManager
import java.util.concurrent.TimeUnit

class MainActivity : AppCompatActivity() {
    lateinit var viewModel:MainVIewModel
    lateinit var plusOneBtn:Button
    lateinit var clearBtn:Button
    lateinit var infoText:TextView
    lateinit var getUserBtn:Button
    lateinit var sp:SharedPreferences

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)
        sp=getPreferences(MODE_PRIVATE)
        val countReserved = sp.getInt("count_reserved",0)
        viewModel = ViewModelProvider(this,MainViewModelFactory(countReserved)).get(MainVIewModel::class.java)
        plusOneBtn = findViewById(R.id.plusOneBtn)
        infoText = findViewById(R.id.infoText)
        clearBtn = findViewById(R.id.clearBtn)
        getUserBtn = findViewById(R.id.getUserBtn)
        plusOneBtn.setOnClickListener {
            viewModel.plusOne()
            refreshCounter()
        }
        clearBtn.setOnClickListener {
            viewModel.clear()
            refreshCounter()
        }
        lifecycle.addObserver(MyObserver(lifecycle))
        viewModel.counter.observe(this, {
            infoText.text = it.toString()
        })

        getUserBtn.setOnClickListener {
            val userId = (0..1000).random().toString()
            viewModel.getUser(userId)
        }
        viewModel.user.observe(this,{
            infoText.text = it.firstName
        })

        //WorkManager
        findViewById<Button>(R.id.doWorkBtn).setOnClickListener {
            val request = OneTimeWorkRequest.Builder(SimpleWorker::class.java)
                .setInitialDelay(5,TimeUnit.SECONDS)
                .setBackoffCriteria(BackoffPolicy.LINEAR,10,TimeUnit.SECONDS)
                .build()
            WorkManager.getInstance(this).enqueue(request)
        }

    }

    override fun onPause() {
        super.onPause()
        sp.edit {
            putInt("count_reserved",viewModel.counter.value?:0)
        }
    }

    private fun refreshCounter(){
        infoText.text = viewModel.counter.toString()
    }
}